File: app/src/main/java/vn/loitp/app/activity/database/sqlitemultitable/SqliteMultiTableActivity.java
package vn.loitp.app.activity.database.sqlitemultitable;

import android.os.Bundle;

import com.core.base.BaseFontActivity;
import com.core.utilities.LLog;

import java.util.List;

import loitp.basemaster.R;
import vn.loitp.app.activity.database.sqlitemultitable.helper.DatabaseHelper;
import vn.loitp.app.activity.database.sqlitemultitable.model.Tag;
import vn.loitp.app.activity.database.sqlitemultitable.model.Note;
import vn.loitp.app.app.LApplication;

//https://www.androidhive.info/2013/09/android-sqlite-database-with-multiple-tables/
public class SqliteMultiTableActivity extends BaseFontActivity {

    DatabaseHelper db;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        test();
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_sqlite_multi_table;
    }

    private void test() {
        db = new DatabaseHelper(getApplicationContext());

        // Creating tags
        Tag tag1 = new Tag("Shopping" + System.currentTimeMillis());
        Tag tag2 = new Tag("Important" + System.currentTimeMillis());
        Tag tag3 = new Tag("Watchlist" + System.currentTimeMillis());
        Tag tag4 = new Tag("Androidhive" + System.currentTimeMillis());

        // Inserting tags in db
        long tag1Id = db.createTag(tag1);
        long tag2Id = db.createTag(tag2);
        long tag3Id = db.createTag(tag3);
        long tag4Id = db.createTag(tag4);

        LLog.d(TAG, "tag1Id: " + tag1Id);
        LLog.d(TAG, "tag2Id: " + tag2Id);
        LLog.d(TAG, "tag3Id: " + tag3Id);
        LLog.d(TAG, "tag4Id: " + tag4Id);

        List<Tag> tagList = db.getTagList();
        LLog.d(TAG, "tagList size: " + tagList.size());
        for (int i = 0; i < tagList.size(); i++) {
            Tag t = tagList.get(i);
            LLog.d(TAG, "tagList -> " + i + " -> " + LApplication.Companion.getGson().toJson(t));
        }

        // Creating ToDos
        Note todo1 = new Note("iPhone 5S", 0);
        Note todo2 = new Note("Galaxy Note II", 0);
        Note todo3 = new Note("Whiteboard", 0);

        Note todo4 = new Note("Riddick", 0);
        Note todo5 = new Note("Prisoners", 0);
        Note todo6 = new Note("The Croods", 0);
        Note todo7 = new Note("Insidious: Chapter 2", 0);

        Note todo8 = new Note("Don't forget to call MOM", 0);
        Note todo9 = new Note("Collect money from John", 0);

        Note todo10 = new Note("Post new Article", 0);
        Note todo11 = new Note("Take database backup", 0);

        // Inserting todos in db
        // Inserting todos under "Shopping" Tag
        long todo1Id = db.createToDo(todo1, new long[]{tag1Id});
        long todo2Id = db.createToDo(todo2, new long[]{tag1Id});
        long todo3Id = db.createToDo(todo3, new long[]{tag1Id});

        // Inserting todos under "Important" Tag
        long todo8Id = db.createToDo(todo8, new long[]{tag2Id});
        long todo9Id = db.createToDo(todo9, new long[]{tag2Id});

        // Inserting todos under "Watchlist" Tag
        long todo4Id = db.createToDo(todo4, new long[]{tag3Id});
        long todo5Id = db.createToDo(todo5, new long[]{tag3Id});
        long todo6Id = db.createToDo(todo6, new long[]{tag3Id});
        long todo7Id = db.createToDo(todo7, new long[]{tag3Id});

        // Inserting todos under "Androidhive" Tag
        long todo10Id = db.createToDo(todo10, new long[]{tag4Id});
        long todo11Id = db.createToDo(todo11, new long[]{tag4Id});

        // "Post new Article" - assigning this under "Important" Tag
        // Now this will have - "Androidhive" and "Important" Tags
        db.createTodoTag(todo10Id, tag2Id);

        // Getting all Todos
        int toDoCount = db.getToDoCount();
        LLog.d(TAG, "toDoCount getToDoCount: " + toDoCount);

        List<Note> todoList = db.getToDoList();
        LLog.d(TAG, "todoList size: " + todoList.size());
        for (int i = 0; i < todoList.size(); i++) {
            Note td = todoList.get(i);
            LLog.d(TAG, ">todoList " + i + " -> " + LApplication.Companion.getGson().toJson(td));
        }

        // Getting todos under "Watchlist" tag name
        List<Note> tagsWatchList = db.getAllToDosByTag(tag3.getTagName());
        for (int i = 0; i < tagsWatchList.size(); i++) {
            Note td = tagsWatchList.get(i);
            LLog.d(TAG, ">tagsWatchList " + i + " -> " + LApplication.Companion.getGson().toJson(td));
        }

        // Deleting
        LLog.d(TAG, "Tag Count Before Deleting: " + db.getToDoCount());
        db.deleteToDo(todo8Id);
        LLog.d(TAG, "Tag Count After Deleting: " + db.getToDoCount());

        // Deleting all Todos under "Shopping" tag
        LLog.d(TAG, "Tag Count Before Deleting 'Shopping' Todos: " + db.getToDoCount());
        db.deleteTag(tag1, true);
        LLog.d(TAG, "Tag Count After Deleting 'Shopping' Todos: " + db.getToDoCount());

        // Updating tag name
        tag3.setTagName("Movies to watch");
        db.updateTag(tag3);

        // Don't forget to close database connection
        db.closeDB();
    }

    @Override
    protected void onDestroy() {
        db.deleteAllDatabase();
        db.closeDB();
        super.onDestroy();
    }
}

File: app/src/main/java/vn/loitp/app/activity/database/sqlitemultitable/SqliteMultiTableActivity.kt
package vn.loitp.app.activity.database.sqlitemultitable

import android.os.Bundle
import android.util.TypedValue
import android.widget.TextView
import com.core.base.BaseFontActivity
import com.core.utilities.LLog
import com.core.utilities.LUIUtil
import kotlinx.android.synthetic.main.activity_sqlite_multi_table.*
import vn.loitp.app.R
import vn.loitp.app.activity.database.sqlitemultitable.helper.DatabaseHelper
import vn.loitp.app.activity.database.sqlitemultitable.model.Note
import vn.loitp.app.activity.database.sqlitemultitable.model.Tag
import vn.loitp.app.app.LApplication

//https://www.androidhive.info/2013/09/android-sqlite-database-with-multiple-tables/
class SqliteMultiTableActivity : BaseFontActivity() {
    private lateinit var db: DatabaseHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        test()
    }

    override fun setFullScreen(): Boolean {
        return false
    }

    override fun setTag(): String? {
        return javaClass.simpleName
    }

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_sqlite_multi_table
    }

    private fun test() {
        db = DatabaseHelper(applicationContext)

        // Creating tags
        val tag1 = Tag("Shopping" + System.currentTimeMillis())
        val tag2 = Tag("Important" + System.currentTimeMillis())
        val tag3 = Tag("Watchlist" + System.currentTimeMillis())
        val tag4 = Tag("Androidhive" + System.currentTimeMillis())

        // Inserting tags in db
        val tag1Id = db.createTag(tag1)
        val tag2Id = db.createTag(tag2)
        val tag3Id = db.createTag(tag3)
        val tag4Id = db.createTag(tag4)

        showMsg("tag1Id: $tag1Id")
        showMsg("tag2Id: $tag2Id")
        showMsg("tag3Id: $tag3Id")
        showMsg("tag4Id: $tag4Id")

        val tagList = db.getTagList()
        showMsg("tagList size: " + tagList.size)
        for (i in tagList.indices) {
            val t = tagList[i]
            showMsg("tagList -> " + i + " -> " + LApplication.gson.toJson(t))
        }

        // Creating note
        val note1 = Note("iPhone 5S", 0)
        val note2 = Note("Galaxy Note II", 0)
        val note3 = Note("Whiteboard", 0)

        val note4 = Note("Riddick", 0)
        val note5 = Note("Prisoners", 0)
        val note6 = Note("The Croods", 0)
        val note7 = Note("Insidious: Chapter 2", 0)

        val note8 = Note("Don't forget to call MOM", 0)
        val note9 = Note("Collect money from John", 0)

        val note10 = Note("Post new Article", 0)
        val note11 = Note("Take database backup", 0)

        // Inserting note in db
        // Inserting note under "Shopping" Tag
        val note1Id = db.createNote(note1, longArrayOf(tag1Id))
        val note2Id = db.createNote(note2, longArrayOf(tag1Id))
        val note3Id = db.createNote(note3, longArrayOf(tag1Id))

        // Inserting note under "Important" Tag
        val note8Id = db.createNote(note8, longArrayOf(tag2Id))
        val note9Id = db.createNote(note9, longArrayOf(tag2Id))

        // Inserting note under "Watchlist" Tag
        val note4Id = db.createNote(note4, longArrayOf(tag3Id))
        val note5Id = db.createNote(note5, longArrayOf(tag3Id))
        val note6Id = db.createNote(note6, longArrayOf(tag3Id))
        val note7Id = db.createNote(note7, longArrayOf(tag3Id))

        // Inserting note under "Androidhive" Tag
        val note10Id = db.createNote(note10, longArrayOf(tag4Id))
        val note11Id = db.createNote(note11, longArrayOf(tag4Id))

        // "Post new Article" - assigning this under "Important" Tag
        // Now this will have - "Androidhive" and "Important" Tags
        db.createNoteTag(note10Id, tag2Id)

        val noteCount = db.getNoteCount()
        showMsg("getNoteCount: $noteCount")

        val noteList = db.getNoteList()
        showMsg("noteList size: " + noteList.size)
        for (i in noteList.indices) {
            val td = noteList[i]
            showMsg(">noteList " + i + " -> " + LApplication.gson.toJson(td))
        }

        // Getting note under "Watchlist" tag name
        val tagsWatchList = db.getAllNoteByTag(tag3.tagName)
        for (i in tagsWatchList.indices) {
            val td = tagsWatchList[i]
            showMsg(">tagsWatchList " + i + " -> " + LApplication.gson.toJson(td))
        }

        // Deleting
        showMsg("Tag Count Before Deleting: " + db.getNoteCount())
        db.deleteNote(note8Id)
        showMsg("Tag Count After Deleting: " + db.getNoteCount())

        // Deleting all note under "Shopping" tag
        showMsg("Tag Count Before Deleting 'Shopping' note: " + db.getNoteCount())
        db.deleteTag(tag1, true)
        showMsg("Tag Count After Deleting 'Shopping' note: " + db.getNoteCount())

        // Updating tag name
        tag3.tagName = "Movies to watch"
        db.updateTag(tag3)

        // Don't forget to close database connection
        db.closeDB()
    }

    override fun onDestroy() {
        db.deleteAllDatabase()
        db.closeDB()
        super.onDestroy()
    }

    private fun showMsg(msg: String) {
        LLog.d(TAG, msg)
        val tv = TextView(activity)
        tv.text = msg
        LUIUtil.setTextSize(tv, TypedValue.COMPLEX_UNIT_DIP, 10)
        ll.addView(tv)
    }
}
